demo/Alsa cpp/main.cpp

```diff
@@ -0,0 +1,216 @@
+snd_pcm_hw_params_t* hw_params;
+snd_pcm_info_t* s_info;
+unsigned int srate = 16000;
+unsigned int nchan = 1;
+
+
+int init_soundcard(const char *devicename)
+{
+    //connect to and set parameters for the sound device
+    //returns any error setting or connecting
+
+	int err = 0;
+
+    if (devicename)
+        strcpy(snd_device, devicename);
+
+	if ((err = snd_pcm_open(&capture_handle, snd_device,SND_PCM_STREAM_CAPTURE, 0)) < 0)
+	{
+		fprintf(stderr, "cannot open audio device %s (%s, %d)\n",snd_device, snd_strerror(err), err);
+		return OPEN_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_malloc(&hw_params)) < 0)
+	{
+		fprintf(stderr,"cannot allocate hardware parameter structure (%s, %d)\n",snd_strerror(err), err);
+		return MALLOC_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_any(capture_handle, hw_params)) < 0)
+	{
+		fprintf(stderr,"cannot initialize hardware parameter structure (%s, %d)\n",snd_strerror(err), err);
+		return ANY_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_set_access(capture_handle, hw_params,SND_PCM_ACCESS_RW_INTERLEAVED)) < 0)
+	{
+		fprintf(stderr, "cannot set access type (%s, %d)\n",snd_strerror(err), err);
+		return ACCESS_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_set_format(capture_handle, hw_params,SND_PCM_FORMAT_S16_LE)) < 0)
+	{
+		fprintf(stderr, "cannot set sample format (%s, %d)\n",snd_strerror(err), err);
+		return FORMAT_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_set_rate_near(capture_handle, hw_params,&srate, 0)) < 0)
+	{
+		fprintf(stderr, "cannot set sample rate (%s, %d)\n",snd_strerror(err), err);
+		return RATE_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params_set_channels(capture_handle, hw_params, nchan))< 0)
+	{
+		fprintf(stderr, "cannot set channel count (%s, %d)\n",snd_strerror(err), err);
+		return CHANNELS_ERROR;
+	}
+
+	if ((err = snd_pcm_hw_params(capture_handle, hw_params)) < 0)
+	{
+		fprintf(stderr, "cannot set parameters (%s, %d)\n",snd_strerror(err), err);
+		return PARAMS_ERROR;
+	}
+
+	if ((err = snd_pcm_prepare(capture_handle)) < 0)
+	{
+		fprintf(stderr,"cannot prepare audio interface for use (%s, %d)\n",snd_strerror(err), err);
+		return PREPARE_ERROR;
+	}
+	if ((err = snd_pcm_start(capture_handle)) < 0)
+	{
+		fprintf(stderr,"cannot start soundcard (%s, %d)\n",snd_strerror(err), err);
+		return START_ERROR;
+	}
+
+	snd_pcm_hw_params_free(hw_params);
+
+	return NO_ERROR;
+}
+
+int close_soundcard()
+{
+    //disconnect for soundcard
+	return snd_pcm_close(capture_handle);
+}
+
+int start_porcupine()
+{
+    //connect to porcupine
+    pv_status_t status = pv_porcupine_init(MODEL_PATH, KEYWORD_PATH, 0.5, &porcupineObject);
+    if (status == PV_STATUS_SUCCESS) {
+        fprintf (stdout, "Porcupine initialized\n");
+    } else {
+        fprintf (stderr, "Failed to initialize Porcupine\n");
+        return START_PORCUPINE_ERROR;
+    }
+    return NO_ERROR;
+}
+
+int listen()
+{
+    //listen on the mic for the wake word
+
+	int err = 0;
+    short wav_data[MAX_BUF_SIZE];
```

Comment:
MAX_BUF_SIZE is really the buffer size. Maybe we can rename it to that? Also, we probably should grab this from pv_porcupine_frame_length() so if we change it at a later point to something like 256 the demo keeps working?